class/practice_2.2/main.cpp

```diff
@@ -0,0 +1,126 @@
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <vector>
+#include <string>
+
+using namespace std;
+
+struct Pixel {
+    int r, g, b;
+
+    bool operator==(const Pixel& other) const {
+        return r == other.r && g == other.g && b == other.b;
+    }
+};
+
+const int WIDTH = 5;
+const int HEIGHT = 5;
+
+bool readImage(const string& filename, vector<vector<Pixel>>& image, const string& outputFilename) {
+    ifstream inFile(filename);
+    ofstream outFile(outputFilename);
+    if (!inFile) {
+        cerr << "Unable to open input file." << endl;
+        return false;
+    }
+
+    string line;
+    int row = 0;
+    while (getline(inFile, line)) {
+        stringstream ss(line);
+        vector<Pixel> pixelRow;
+        for (int col = 0; col < WIDTH; ++col) {
+            Pixel pixel;
+            char comma;
+            ss >> pixel.r >> comma >> pixel.g >> comma >> pixel.b;
+            if (comma != ',' || pixel.r < 0 || pixel.r > 255 || pixel.g < 0 || pixel.g > 255 || pixel.b < 0 || pixel.b > 255) {
+                cerr << "Invalid input format at row " << row + 1 << ", column " << col + 1 << "." << endl;
+                return false;
+            }
+            pixelRow.push_back(pixel);
```

Comment:
even if you will use vector instead of array then it is reasonable to preallocate vector with a given size to avoid unneeded memory reallocation. Of course you will need to adjust loop accordingly to handle the case of extra data in file but it should not be a big problem.